trikot-streams/streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/promise/Promise.kt

```diff
@@ -20,50 +22,71 @@ class Promise<T> internal constructor(
     private val serialQueue = SynchronousSerialQueue()
     private val result = BehaviorSubjectImpl<T>(serialQueue = serialQueue)
 
-    private val isCancelled: AtomicReference<Boolean> = AtomicReference(false)
-    private val internalCancellableManager: CancellableManager = CancellableManager().also {
-        cancellableManager?.add(it)
+    private val internalCancellableManager: CancellableManager = CancellableManager()
+
+    private val onParentCancellation = object : Cancellable, VerifiableCancelledState {
+        override var isCancelled: Boolean by atomic(false)
+
+        override fun cancel() {
+            serialQueue.dispatch {
+                isCancelled = true
+                internalCancellableManager.cancel()
+
+                if (result.value == null && result.error == null) {
+                    result.error = CancelledPromiseException
+                }
+            }
+        }
+    }
+
+    private val isParentCancellableCancelled: Boolean
+        get() = onParentCancellation.isCancelled
+
+    /**
+     * When a result is received, we want to make sure we're considered as "cancelled" by the provided cancellableManager if any.
+     * This allows to use `CancellableManager::cleanCancelledChildren` with promises to clean up finished promises
+     * and avoids keeping a reference on the promise, which is finished anyway. No further results will be emitted.
+     *
+     * We don't want the `onParentCancellation::cancel()` to run since the provided cancellableManager was not cancelled,
+     * so we don't call `cancel()` directly, but instead modify the `isCancelled` value manually.
+      */
+    private fun onResultReceived() {
```

Comment:
Is there any way we can test that in https://github.com/mirego/trikot/blob/master/trikot-streams/streams/src/commonTest/kotlin/com/mirego/trikot/streams/reactive/promise/PromiseCancellationTests.kt ?

Reply:
`onResultReceived` only affect internal stuff that are not accessible outside of the Promise.

Unfortunately I don't think this part is testable via unit test unless we make some attributes public only for testing purposes, which I don't think is ideal.